api/src/main/java/org/apache/iceberg/stats/StatsUtil.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.stats;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.types.Types;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StatsUtil {
+  private static final Logger LOG = LoggerFactory.getLogger(StatsUtil.class);
+  static final int VALUE_COUNT_OFFSET = 0;
+  static final int NULL_VALUE_COUNT_OFFSET = 1;
+  static final int NAN_VALUE_COUNT_OFFSET = 2;
+  static final int AVG_VALUE_SIZE_OFFSET = 3;
+  static final int MAX_VALUE_SIZE_OFFSET = 4;
+  static final int LOWER_BOUND_OFFSET = 5;
+  static final int UPPER_BOUND_OFFSET = 6;
+  static final int NUM_STATS_PER_COLUMN = 200;
+  static final int RESERVED_FIELD_IDS = 200;
+  static final int DATA_SPACE_FIELD_ID_START = 10_000;
+  static final int METADATA_SPACE_FIELD_ID_START = 2_147_000_000;
+  static final int RESERVED_FIELD_IDS_START = Integer.MAX_VALUE - RESERVED_FIELD_IDS;
+  // support stats for only up to this amount of data field IDs
+  static final int MAX_DATA_FIELD_ID = 1_000_000;
+  static final int MAX_DATA_STATS_FIELD_ID = 200_010_000;
+
+  private StatsUtil() {}
+
+  public static int statsFieldIdForField(int fieldId) {
```

Comment:
nit: statsFieldStartingIdForDataField(init fieldId)? this is the starting/base id